rigid_opt/sdf_2_sdf_optimizer2d.py

```diff
@@ -0,0 +1,138 @@
+        :param visualization_parameters:
+        """
+
+        if verbosity_parameters:
+            self.verbosity_parameters = verbosity_parameters
+        else:
+            self.verbosity_parameters = Sdf2SdfOptimizer2d.VerbosityParameters()
+
+        if visualization_parameters:
+            self.visualization_parameters = visualization_parameters
+        else:
+            self.visualization_parameters = Sdf2SdfVisualizer.Parameters()
+
+        self.visualizer = None
+
+    def optimize(self,
+                 data_to_use,
+                 eta=.01,
+                 voxel_size=0.004,
+                 narrow_band_width_voxels=20.,
+                 iteration=60,
+                 ):
+        """
+        Optimization algorithm
+        :param data_to_use:
+        :param eta: thickness of surface, used to determine reliability of sdf field
+        :param iteration: total number of iterations
+        :param voxel_size: voxel side length
+        :param narrow_band_width_voxels:
+        :return:
+        """
+
+        canonical_field = data_to_use.generate_2d_canonical_field(narrow_band_width_voxels=narrow_band_width_voxels,
+                                                                  method=tsdf_gen.DepthInterpolationMethod.NONE)
+        live_field = data_to_use.generate_2d_live_field(narrow_band_width_voxels=narrow_band_width_voxels,
+                                                        method=tsdf_gen.DepthInterpolationMethod.NONE)
+        field_size = canonical_field.shape[0]
+        offset = data_to_use.offset
+        twist = np.zeros((3, 1))
+
+        self.visualizer = Sdf2SdfVisualizer(parameters=self.visualization_parameters, field_size=field_size)
+        self.visualizer.generate_pre_optimization_visualizations(canonical_field, live_field)
+
+        for iteration_count in range(iteration):
+            matrix_a = np.zeros((3, 3))
+            vector_b = np.zeros((3, 1))
+            canonical_weight = (canonical_field > -eta).astype(np.int)
+            twist_3d = np.array([twist[0],
+                                 [0.],
+                                 twist[1],
+                                 [0.],
+                                 twist[2],
+                                 [0.]], dtype=np.float32)
+            live_field = data_to_use.generate_2d_live_field(narrow_band_width_voxels=narrow_band_width_voxels,
+                                                            method=tsdf_gen.DepthInterpolationMethod.NONE,
+                                                            apply_transformation=True, twist=twist_3d)
+            live_weight = (live_field > -eta).astype(np.int)
+            live_gradient = GradientField().calculate(live_field, twist, array_offset=offset, voxel_size=voxel_size)
+
+            for i in range(live_field.shape[0]):
+                for j in range(live_field.shape[1]):
+                    matrix_a += np.dot(live_gradient[i, j][:, None], live_gradient[i, j][None, :])
```

Comment:
Could you explain what exactly this accesses?
`live_gradient[i, j][:, None]`

Reply:
It converts the shape of `live_gradient[i, j]` from (3,) to (3 ,1), so `np.dot()` works correctly

Reply:
I see, thank you. I thought about this, vectorization doesn't seem trivial, let's leave this as is.